### PR TITLE
[mDNS] mDNS implementation using mojo interface

### PR DESCRIPTION
**Pull request by gjc13**
On chromecast devices, the mdns is implemented by the chromecast application. This change calls the mojo RPC interface to publish services.

### CODE REVIEW

**Comment by gjc13** [comment]
The CI test scripts are still to be added.

File: src/agent/border_agent.cpp [hunk 1]
```diff
@@ -77,7 +77,7 @@ enum
 };
 
 BorderAgent::BorderAgent(Ncp::Controller *aNcp)
-#if OTBR_ENABLE_MDNS_AVAHI || OTBR_ENABLE_MDNS_MDNSSD
+#if OTBR_ENABLE_MDNS_AVAHI || OTBR_ENABLE_MDNS_MDNSSD || OTBR_ENABLE_MOJO
```
Comment (bukepo) [inline src/agent/border_agent.cpp]: `OTBR_ENABLE_MOJO` -> `OTBR_ENABLE_MDNS_MOJO`?

File: src/agent/mdns_mojo.cpp [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+void MdnsMojoPublisher::PublishServiceTask(uint16_t                        aPort,
+                                           const std::string &             aType,
+                                           const std::string &             aInstanceName,
+                                           const std::vector<std::string> &aText)
+{
+    std::string            serviceName;
+    std::string            serviceProtocol;
+    std::string::size_type split = aType.rfind('.');
```
Comment (bukepo) [inline src/agent/mdns_mojo.cpp]: is it possible to simplify to:
```
std::string::size_type split = aType.rfind('.', aType.length() - 2);
```
line 156-159 can be saved.

File: src/agent/mdns_mojo.cpp [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    {
+        split = aType.rfind('.', split - 1);
+    }
+
+    VerifyOrExit(split != std::string::npos);
+
+    serviceName     = aType.substr(0, split);
+    serviceProtocol = aType.substr(split + 1, std::string::npos);
+
+    VerifyOrExit(!serviceName.empty() && !serviceProtocol.empty());
```
Comment (bukepo) [inline src/agent/mdns_mojo.cpp]: move this after removing the trailing dot.

File: src/agent/mdns_mojo.cpp [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+#include <functional>
+
+#include "common/code_utils.hpp"
+#ifndef TEST_IN_CHROMIUM
+#include "agent/mdns_mojo.hpp"
+#include "common/logging.hpp"
+#else
+#include "mdns_mojo.hpp"
+#define otbrLog(level, format, ...) fprintf(stderr, format, ##__VA_ARGS__)
```
Comment (bukepo) [inline src/agent/mdns_mojo.cpp]: define as the following?
```cpp
#define otbrLog(level, format, ...) do {\
fprintf(stderr, format, ##__VA_ARGS__);\
fprintf(stderr, "\r\n");\
}
```

File: src/agent/mdns_mojo.hpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
+/*
+ *    Copyright (c) 2017, The OpenThread Authors.
```
Comment (bukepo) [inline src/agent/mdns_mojo.hpp]: -> `2019`

File: public/chromecast/mojom/mdns.mojom [hunk 6]
```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
```
Comment (bukepo) [inline public/chromecast/mojom/mdns.mojom]: `2019`?

File: tests/mdns/mdns_mojo_test/stub_mojo_service.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
+#include <base/at_exit.h>
```
Comment (bukepo) [inline tests/mdns/mdns_mojo_test/stub_mojo_service.cpp]: missing license.

File: tests/mdns/test-mojo [hunk 8]
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env bash
```
Comment (bukepo) [inline tests/mdns/test-mojo]: just /usr/bash?

File: tests/mdns/mdns_mojo_test/stub_mojo_service.cpp [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+#include <base/message_loop/message_pump_for_io.h>
+#include <base/run_loop.h>
+#include <chromecast/external_mojo/external_service_support/external_connector.h>
+#include <chromecast/external_mojo/external_service_support/external_service.h>
+#include <chromecast/external_mojo/public/cpp/common.h>
+#include <mojo/core/embedder/embedder.h>
+#include <mojo/core/embedder/scoped_ipc_support.h>
+#include <mojo/public/cpp/bindings/binding_set.h>
+
+#include <cstdio>
```
Comment (bukepo) [inline tests/mdns/mdns_mojo_test/stub_mojo_service.cpp]: what's this for?

File: src/agent/mdns_mojo.hpp [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    bool IsStarted(void) const override;
+
+    /**
+     * This method publishes or updates a service.
+     *
+     * @param[in]   aName               The name of this service.
+     * @param[in]   aType               The type of this service.
+     * @param[in]   aPort               The port number of this service.
+     * @param[in]   ...                 Pointers to null-terminated string of key
+     * and value for text record. The last argument must be NULL.
```
Comment (bukepo) [inline src/agent/mdns_mojo.hpp]: indent to right?

File: src/agent/mdns_mojo.hpp [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+#ifndef TEST_IN_CHROMIUM
+#include <base/task_scheduler/post_task.h>
+#else
+#include <base/task/post_task.h>
+#endif
+#include <chromecast/external_mojo/public/cpp/common.h>
+#ifndef TEST_IN_CHROMIUM
+#include <chromecast/external_mojo/public/cpp/external_connector.h>
+#else
+#include <chromecast/external_mojo/external_service_support/external_connector.h>
```
Comment (bukepo) [inline src/agent/mdns_mojo.hpp]: can we put them in the same directory?

File: src/agent/mdns_mojo.cpp (hunk 2 above)
Comment (gjc13) [reply src/agent/mdns_mojo.cpp]: Required to handle aType=="a" or aType=="." corner cases

File: src/agent/mdns_mojo.hpp (hunk 11 above)
Comment (gjc13) [reply src/agent/mdns_mojo.hpp]: These are files in libchrome and chrominum which we cannot modify.

File: tests/mdns/test-mojo [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    cp -r "${OTBR_DIR}/tests/mdns/mdns_mojo_test" "${CHROMIUM_DIR}/src/"
+    ln -s "${OTBR_DIR}/public/chromecast/mojom/mdns.mojom" "${MOJO_TEST_SRC_DIR}/public/mojom/mdns.mojom"
+    ln -s "${OTBR_DIR}/src/agent/mdns.hpp" "${MOJO_TEST_SRC_DIR}/mdns.hpp"
+    ln -s "${OTBR_DIR}/src/agent/mdns_mojo.hpp" "${MOJO_TEST_SRC_DIR}/mdns_mojo.hpp"
+    ln -s "${OTBR_DIR}/src/agent/mdns_mojo.cpp" "${MOJO_TEST_SRC_DIR}/mdns_mojo.cpp"
+    ln -s "${OTBR_DIR}/src/common" "${MOJO_TEST_SRC_DIR}/common"
+
+    cd /etc/apt/sources.list.d && sudo rm -rf cassandra.list* couchdb.list* mongodb-3.4.list* rabbitmq_rabbitmq-server.list* chris-lea-redis-server.list* github_git-lfs.list*
+    cd "${CHROMIUM_DIR}/src"
+    # Required to install sperately on travis
```
Comment (bukepo) [inline tests/mdns/test-mojo]: typo: sperately

**Comment by gjc13** [comment]
@jwhui , friendly reminder

File: public/chromecast/mojom/mdns.mojom [hunk 13]
```diff
@@ -0,0 +1,113 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (jwhui) [inline public/chromecast/mojom/mdns.mojom]: Can we move this to `third_party`?

File: Android.mk [hunk 14]
```diff
@@ -26,12 +26,49 @@
 #  POSSIBILITY OF SUCH DAMAGE.
 #
 
-LOCAL_PATH:= $(call my-dir)
+LOCAL_PATH := $(call my-dir)
```
Comment (bukepo) [inline Android.mk]: Add default with mdns value for Android.mk?

`WITH_MDNS ?= mojo`

Comment (gjc13) [reply Android.mk]: We have tests with no WITH_MDNS specified.

File: tests/mdns/test-mojo [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+# Test build and run mdns using mojo
+#
+
+set -x
+set -e
+
+
+main() 
+{
+    OTBR_DIR=${OTBR_DIR:-$(pwd)}
```
Comment (bukepo) [inline tests/mdns/test-mojo]: Just realized this check seems to be unable to guarantee build with Android.mk when WITH_MDNS=mojo, which I think is one of our main goals for the CI test. And I didn't find other checks verifies this.

File: Android.mk (hunk 14 above)
Comment (bukepo) [reply Android.mk]: WITH_MDNS is used by Android.mk and travis test scripts. It has no default value in Android.mk because we didn't support it before. Now we are adding the support, we should set default value to mojo, thus it can be used without any modifications.

File: tests/mdns/test-mojo (hunk 15 above)
Comment (gjc13) [reply tests/mdns/test-mojo]: The open source version of libchrome does not support mojo external client so maybe we should use mDNSResponder instead?
